obsreport/obsreport_exporter.go

```diff
@@ -0,0 +1,216 @@
+		exporterPrefix+RefusedMetricPointsKey,
+		"Number of metric points in failed attempts to send to destination.",
+		stats.UnitDimensionless)
+)
+
+// StartTraceDataExportOp is called at the start of an Export operation.
+// The returned context should be used in other calls to the obsreport functions
+// dealing with the same export operation.
+func StartTraceDataExportOp(ctx context.Context, exporter string) (context.Context, *trace.Span) {
+	return traceExportDataOp(
+		exporterContext(ctx, exporter),
+		exporter,
+		exportTraceDataOperationSuffix)
+}
+
+// EndTraceDataExportOp completes the export operation that was started with
+// StartTraceDataExportOp.
+func EndTraceDataExportOp(
+	ctx context.Context,
+	span *trace.Span,
+	numExportedSpans int,
+	numDroppedSpans int, // TODO: For legacy measurements, to be removed in the future.
+	err error,
+) {
+	if useLegacy {
+		observability.RecordMetricsForTraceExporter(
+			ctx, numExportedSpans, numDroppedSpans)
+	}
+
+	endExportOp(
+		ctx,
+		span,
+		numExportedSpans,
+		err,
+		traceData,
+	)
+}
+
+// StartMetricsExportOp is called at the start of an Export operation.
+// The returned context should be used in other calls to the obsreport functions
+// dealing with the same export operation.
+func StartMetricsExportOp(
+	ctx context.Context,
+	exporter string,
+) (context.Context, *trace.Span) {
+	return traceExportDataOp(
+		exporterContext(ctx, exporter),
+		exporter,
+		exportMetricsOperationSuffix)
+}
+
+// EndMetricsExportOp completes the export operation that was started with
+// StartMetricsExportOp.
+func EndMetricsExportOp(
+	ctx context.Context,
+	span *trace.Span,
+	numExportedPoints int,
+	numExportedTimeSeries int, // TODO: For legacy measurements, to be removed in the future.
+	numDroppedTimeSeries int, // TODO: For legacy measurements, to be removed in the future.
+	err error,
+) {
+	if useLegacy {
+		observability.RecordMetricsForMetricsExporter(
+			ctx, numExportedTimeSeries, numDroppedTimeSeries)
+	}
+
+	endExportOp(
+		ctx,
+		span,
+		numExportedPoints,
+		err,
+		metricsData,
+	)
+}
+
+// exporterContext adds the keys used when recording observability metrics to
+// the given context returning the newly created context. This context should
+// be used in related calls to the obsreport functions so metrics are properly
+// recorded.
+func exporterContext(
+	ctx context.Context,
+	exporter string,
+) context.Context {
+	if useLegacy {
+		ctx = observability.ContextWithExporterName(ctx, exporter)
+	}
+
+	if useNew {
+		ctx, _ = tag.New(ctx, tag.Upsert(
+			tagKeyExporter, exporter, tag.WithTTL(tag.TTLNoPropagation)))
+	}
+
+	return ctx
+}
+
+// traceExportDataOp creates the span used to trace the operation. Returning
+// the updated context and the created span.
+func traceExportDataOp(
+	exporterCtx context.Context,
+	exporterName string,
+	operationSuffix string,
+) (context.Context, *trace.Span) {
+	spanName := exporterPrefix + exporterName + operationSuffix
```

Comment:
Is it possible to precalculate the spanName once instead of on every export operation?

Reply:
I considered creating an "operation ticket" type to carry this type of information but didn't explore that much. I can give a second look together with the other optimization about metric tags.